File: CallbackTechniques/Delegate/Delegate.m
//
//  Delegate.m
//  CallbackTechniques
//
//  Created by Alexey Naumov on 29/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

#import "Delegate.h"

@implementation ObjCDelegateCaller

- (void)callDelegate
{
  [self.delegate performDelegatedJob];
}

@end

@implementation ObjCDelegateCallee

- (void)performDelegatedJob
{
  self.wasCalled = YES;
}

@end

File: CallbackTechniques/Delegate/Delegate.swift
//
//  Delegate.swift
//  CallbackTechniques
//
//  Created by Alexey Naumov on 29/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

protocol SwiftDelegateCallerDelegate {
  func performDelegatedJob()
}

class SwiftDelegateCaller {
  var delegate: SwiftDelegateCallerDelegate?
  
  func callDelegate() {
    delegate?.performDelegatedJob()
  }
}

class SwiftDelegateCallee: SwiftDelegateCallerDelegate {
  
  var wasCalled = false
  
  func performDelegatedJob() {
    wasCalled = true
  }
}
